corehq/apps/data_analytics/tests/test_malt_generator.py

```diff
@@ -88,12 +86,19 @@ def _save_form_data(self, app_id, received_on):
             app_id=app_id,
         )
 
-    def test_successful_malt_generation(self):
+    def setUp(self) -> None:
+        super().setUp()
+        initialize_index_and_mapping(self.es, XFORM_INDEX_INFO)
+
+    def tearDown(self) -> None:
+        ensure_index_deleted(XFORM_INDEX_INFO.index)
+        super().tearDown()
```

Comment:
These tests are pretty slow, likely because of this. I see some other spots in HQ that have better methods on deleting specific form submissions, but the setup is a bit different from how these forms are created. Rather than try to figure out which way is the best approach, I figured I'd ask reviewers if there is a standard way to go about creating form submissions in es for tests?

Reply:
Heh so I made this change locally to remember the form_ids that were created and delete them at the end of each test but that was even slower. I lost hope in ES and went back to one test to serve as _the_ end to end test for the malt, but added some additional assertions/tests in the single test method which isn't ideal, but runs in a third of the time. See 5cbacc095b25c880782440781e3f3dfd1a80e6a3

Reply:
> `TransactionTestCase` is really slow

Yes, the naming of that class is confusing. It does not use transactions (unlike `TestCase`), but instead allows writing tests that need to see the effects of transactions in the code under test. The way that `TransactionTestCase` resets the database is the real issue.

> TransactionTestCase and TestCase are identical except for the manner in which the database is reset to a known state and the ability for test code to test the effects of commit and rollback:
>
> -   A TransactionTestCase resets the database after the test runs by truncating all tables. A TransactionTestCase may call commit and rollback and observe the effects of these calls on the database.
> -   A TestCase, on the other hand, does not truncate tables after a test. Instead, it encloses the test code in a database transaction that is rolled back at the end of the test. This guarantees that the rollback at the end of the test restores the database to its initial state.

https://docs.djangoproject.com/en/4.0/topics/testing/tools/#transactiontestcase

Reply:
Yeah my understanding is that due to this method:
```
def _save_malt_row_dicts_to_db(malt_row_dicts):
    try:
        MALTRow.objects.bulk_create(
            [MALTRow(**malt_dict) for malt_dict in malt_row_dicts]
        )
    except IntegrityError:
        for malt_row_dict in malt_row_dicts:
            _update_or_create_malt_row(malt_row_dict)
```
Once that exception is thrown, it corrupts the transaction in which this is running, meaning no other changes can be made to the db until this transaction is rolled back? In practice, I get this error: `An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.` Are you aware of a better way around this type of issue?